filebeat/input/mqtt/input.go

```diff
@@ -0,0 +1,194 @@
+	onMessageHandler := createOnMessageHandler(logger, out, inflightMessages)
+	onConnectHandler := createOnConnectHandler(logger, &inputContext, onMessageHandler, clientSubscriptions, newBackoff)
+	clientOptions, err := createClientOptions(config, onConnectHandler)
+	if err != nil {
+		return nil, err
+	}
+
+	return &mqttInput{
+		client:           newMqttClient(clientOptions),
+		inflightMessages: inflightMessages,
+		logger:           logp.NewLogger("mqtt input").With("hosts", config.Hosts),
+	}, nil
+}
+
+func createOnMessageHandler(logger *logp.Logger, outlet channel.Outleter, inflightMessages *sync.WaitGroup) func(client libmqtt.Client, message libmqtt.Message) {
+	return func(client libmqtt.Client, message libmqtt.Message) {
+		inflightMessages.Add(1)
+
+		logger.Debugf("Received message on topic '%s', messageID: %d, size: %d", message.Topic(),
+			message.MessageID(), len(message.Payload()))
+
+		mqttFields := common.MapStr{
+			"duplicate":  message.Duplicate(),
+			"message_id": message.MessageID(),
+			"qos":        message.Qos(),
+			"retained":   message.Retained(),
+			"topic":      message.Topic(),
+		}
+		outlet.OnEvent(beat.Event{
+			Timestamp: time.Now(),
+			Fields: common.MapStr{
+				"message": string(message.Payload()),
+				"mqtt":    mqttFields,
+			},
+		})
+
+		inflightMessages.Done()
+	}
+}
+
+func createOnConnectHandler(logger *logp.Logger,
+	inputContext *input.Context,
+	onMessageHandler func(client libmqtt.Client, message libmqtt.Message),
+	clientSubscriptions map[string]byte,
+	newBackoff func(done <-chan struct{}, init, max time.Duration) backoff.Backoff) func(client libmqtt.Client) {
+	// The function subscribes the client to the specific topics (with retry backoff in case of failure).
+	return func(client libmqtt.Client) {
+		backoff := newBackoff(
+			inputContext.Done,
+			subscribeRetryInterval,
+			8*subscribeRetryInterval)
+
+		var topics []string
+		for topic := range clientSubscriptions {
+			topics = append(topics, topic)
+		}
+
+		var success bool
+		for !success {
+			logger.Debugf("Try subscribe to topics: %v", strings.Join(topics, ", "))
+
+			token := client.SubscribeMultiple(clientSubscriptions, onMessageHandler)
+			if !token.WaitTimeout(subscribeTimeout) {
+				if token.Error() != nil {
+					logger.Warnf("Subscribing to topics failed due to error: %v", token.Error())
+				}
+
+				if !backoff.Wait() {
+					backoff.Reset()
+					success = true
+				}
+			} else {
+				backoff.Reset()
+				success = true
+			}
+		}
+	}
+}
+
+// Run method starts the mqtt input and processing.
+// The mqtt client starts in auto-connect mode (with connection retries and resuming topic subscriptions).
+func (mi *mqttInput) Run() {
+	mi.once.Do(func() {
+		mi.logger.Debug("Run the input once.")
+		mi.client.Connect()
+	})
+}
+
+// Stop method stops the input.
+func (mi *mqttInput) Stop() {
+	mi.logger.Debug("Stop the input.")
+	mi.client.Disconnect(uint(disconnectTimeout.Milliseconds()))
```

Comment:
This could be up to a 3 second delay and `Stop` isn't supposed to block, so `Disconnect` should be called asynchronously (in a goroutine) in a way that can be waited on in `Wait` (probably requires adding another `WaitGroup` or similar in `mqttInput`)

Reply:
Nice spotting! Thank you for bringing this up. Fixed.